### ISSUE TITLE
Invalid Signature on Debian Binaries for 1.0.11

### ISSUE THREAD

**Issue by gesker** [zcash/zcash#2590]
Describe the issue

Bad Debian apt signature

Can you reliably reproduce the issue?

If so, please list the steps to reproduce below:

apt update

Expected behaviour

No signature errors

Actual behaviour + errors

Get: 1 https://apt.z.cash jessie Release.gpg [819 B]
Err https://apt.z.cash jessie Release.gpg
The following signatures were invalid: F1E214037E94E950BA8577B263C4A2169C1B2FA2
Fetched 819 B in 2min 0s (6 B/s)
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://apt.z.cash jessie Release: The following signatures were invalid: F1E214037E94E950BA8577B263C4A2169C1B2FA2
W: Failed to fetch https://apt.z.cash/dists/jessie/Release.gpg: The following signatures were invalid: F1E214037E94E950BA8577B263C4A2169C1B2FA2
W: Some index files failed to download. They have been ignored, or old ones used instead.

The version of Zcash you were using:

1.0.10 trying to go to 1.0.11

Machine specs:

OS name + version: ubuntu 17.04

**Comment by trou** [comment]
More info:
```
apt-get -o Debug::Acquire::gpgv=true update
[...]
Get:13 https://apt.z.cash jessie Release [2288 B]                                  
Get:14 https://apt.z.cash jessie Release.gpg [819 B]                                        
0% [9 Sources store 39.9 MB] [12 Packages 572 kB/7601 kB 8%]inside VerifyGetSigners           
0% [9 Sources store 39.9 MB] [12 Packages 572 kB/7601 kB 8%] [14 Release.gpg gpgv 2288 B]Preparing to exec:  /usr/bin/apt-key --quiet --readonly verify --status-fd 3 /var/lib/apt/lists/partial/apt.z.cash_dists_jessie_Release.gpg /var/lib/apt/lists/partial/apt.z.cash_dists_jessie_Release
0% [12 Packages 662 kB/7601 kB 9%] [14 Release.gpg gpgv 2288 B]Read: [GNUPG:] NEWSIG     

Read: [GNUPG:] KEY_CONSIDERED F1E214037E94E950BA8577B263C4A2169C1B2FA2 0

Read: [GNUPG:] SIG_ID B1BVlGCUSwuR3HMNHhdhO4d6+Lw 2017-08-16 1502843557

Read: [GNUPG:] KEY_CONSIDERED F1E214037E94E950BA8577B263C4A2169C1B2FA2 0

Read: [GNUPG:] GOODSIG 63C4A2169C1B2FA2 Zcash Master Signing Key

Got GOODSIG 63C4A2169C1B2FA2 !
Read: [GNUPG:] VALIDSIG F1E214037E94E950BA8577B263C4A2169C1B2FA2 2017-08-16 1502843557 0 4 0 1 2 00 F1E214037E94E950BA8577B263C4A2169C1B2FA2

Got untrusted VALIDSIG, key ID: F1E214037E94E950BA8577B263C4A2169C1B2FA2
gpgv exited with status 0
[...]
```
Looking at apt source code:

https://github.com/Debian/apt/blob/master/methods/gpgv.cc#L22:

it seems clear SHA-1's the culprit.

**Comment by str4d** [comment]
Thanks for investigating this!

The relevant commit that deprecated SHA1 in APT is https://github.com/Debian/apt/commit/33d7a8d672c8c720947e81158de4a5a07be05b72. This clarifies why the 1.0.11 signature is failing on newer Debian-based OSs.

The next question is why a SHA1 signature was generated. Previous releases have clearly had signatures trusted by these OSs, which suggests that this was caused by differences in the local environments of the package signers. #2572 is about generating a Debian package as part of our deterministic builds, but it still intends for user-accepting signatures to be created manually by a package signer, so these differences will continue to matter.

Once we've figured out what caused the signature to use SHA1, we should create a script for signing packages that ensures the signatures are compatible with all supported OSs. Testing the packages in CI would also be useful (#1562).

**Comment by trou** [comment]
It's probably a matter of either:
* using a recent version of GPG with SHA-256 as default
* making sure the default hash algorithm is SHA-256 (see https://debian-administration.org/users/dkg/weblog/48)
* explicitly passing the hash algorithm to GPG

**Comment by interfect** [comment]
So what's the recommended installation method to use while the repository is down? Are there individually-signed debs? Or should everyone build from source? (And should the web site be updated to reflect that?)

**Comment by arcalinea** [comment]
@str4d Do you know if this issue is still occurring after the tarball was re-signed?

**Comment by webwurst** [comment]
I just tried again and during installation no error about invalid signatures showed up. So I guess it's fixed. Thanks!

**Comment by daira** [comment]
This can be closed, since #1562 is about testing that would pick up any failure to sign with the correct algorithm.
